File: PHP/markov.php
<?php
/**
 * Used to generate markov-chain-based text
 */
class Markov{
	// Create all of our private variables
	private $store;
	private $storeSize;
	private $file;
	private $words;
	private $wordsSize;

	/**
	 * @param String $file The contents of a text file
	 */
	function __construct( $file ) {
		// Setup our file and store, then process the file's contents
		// and create a simple probability array
		$this->file = $file;
		$this->store = [];
		$this->words = $this->splitWords( );
		$this->wordsSize = count( $this->words );
		$this->storeProbability( );
		$this->storeSize = count( $this->store );
	}

	/**
	 * Splits the string into an array of all the words delimited by a space
	 * @return Array
	 */
	private function splitWords( ){
		$wordSplit = explode(' ', $this->file);
		return $wordSplit;
	}

	/**
	 * Creates a simple 2nd Order probability table for the words
	 * @return Array
	 */
	private function storeProbability( ){
		if( $this->wordsSize < 3 ){
			// If the total number of words is less than 3, we can't use this text
			return;	
		} else {
			for ($i=0; $i < ($this->wordsSize - 2); $i++) { 
				// We'll use a comma-separated string as key 
				// and explode it later rather than messing 
				// with the SplObjectStorage class
				$key = $this->words[ $i ] .','. $this->words[ $i + 1 ];
				$this->store[ $key ][] = $this->words[ $i + 2 ];
			}
		}
	}

	/**
	 * Generate the markov chain text based on the calculated store
	 * @param  integer $wordLimit The amount of words to be returned
	 * @return String             The resulting text
	 */
	public function generateMarkovText( $wordLimit = 250 ){
		$output = '';

		if( !empty( $this->store ) ){
			// If store has more than three words
			
			// Grab a random key from our array
			$key = array_rand( $this->store );
			for ($i=0; $i < $wordLimit; $i++) {
				// For until we reach the word limit

				// Extract our individual words from the key and output the first
				$words = explode(',', $key);
				$output .= ' '.$words[ 0 ];

				if( isset( $this->store[ $key ] ) ){
					// If the new key exists

					// Randomly select one of the new words and get a new key
					$randomInteger = mt_rand( 0, count( $this->store[$key] ) - 1 );
					$key = $words[ 1 ].','.$this->store[ $key ][ $randomInteger ];
				} else {
					// If it doesn't exist, grab another random one
					$key = array_rand( $this->store );
				}
			}
			// Bit of beautification for the output string
			return ucfirst( trim( $output ) ).'.';
		}
		// If the store's empty, a text with less than 3 words was supplied
		return 'Please use a text with more than 3 words.';
	}
}

?>

File: Java/Markov.java
import java.util.*;
import java.io.*;
import java.util.Random;

/**
 * Generates text based on a Markov Chain
 */
public class Markov{
	public static void main( String[] args ){
		Markov markov = new Markov( );
		System.out.println( markov.generateMarkovText( 250 ) );
	}

	private final String[] words;
	private final HashMap<String, ArrayList<String>> store;
	private final String file;
	private final Object[] keys;

	/**
	 * Opens a file and calculates the probability of words 
	 */
	public Markov( ){
		this.store = new HashMap<String, ArrayList<String>>( );
		String content = "";

		try{
			content = new Scanner(new File("../test_text.txt")).useDelimiter("\\Z").next();
		} catch (FileNotFoundException e) {
			// If file not found, let someone know and exit out
		    System.out.println( "The specified file was not found" );
		    e.printStackTrace();
		    System.exit( 1 );
		}

		// Put the file's contents into an instance variable
		this.file = content;
		// Split the contents into an array of words
		this.words = content.split(" ");
		this.storeProbablity( );
		this.keys = this.store.keySet( ).toArray( );
	}

	/**
	 * Calculates the probability of succeeding words in a text
	 */
	public void storeProbablity( ){
		int wordSize = this.words.length;
		if( wordSize < 3 ){
			// If the text is too small, error out
			System.out.println( "Please use a larger sample text" );
			System.exit( 1 );
		} else {
			for (int i = 0; i < (wordSize-2); i++) {
				// For each triple of words
				String key = this.words[i] + ">>" + this.words[i+1];

				if( ! ( this.store.containsKey( key ) ) ){
					// If the store doesn't contain the key already, initialise it
					// as an empty ArrayList
					this.store.put( key, new ArrayList<String>( ) );
				}
				// Add the succeeding word to the store
				this.store.get( key ).add( this.words[i+2] );
			}
		}
	}

	/**
	 * Generates text based on a Markov Chain
	 * @param  wordLimit The number of words to be output
	 * @return           The generated text
	 */
	public String generateMarkovText( int wordLimit ){
		String output = "";

		// Get a random key to begin with
		Random random = new Random( );
		int randomInt = random.nextInt( this.store.size( ) - 1 );
		String key = (String) this.keys[ randomInt ];
		
		for (int i=0; i < wordLimit; i++) {

			// Split the words into individual words
			String[] words = key.split(">>");

			output += words[0] + " ";

			// The number of possibilities of the size of a key's array
			int possibilities = this.store.get( key ).size( );
			if( possibilities < 1 ){
				// If there is more than 1 possibility, randomly get a succeeding word
				// and use that for the next key
				int randomPossibility = random.nextInt( this.store.size( ) - 1 );
				key = words[1] + ">>" + this.store.get( key ).get( randomPossibility );
			} else {
				// If there are no possibilities, randomly get another key
				int randomPossibility = random.nextInt( this.store.size( ) - 1 );
				key = (String) this.keys[ randomPossibility ];
			}
		}
		return output.replace( "\n", " " );
	}
}